java/algorithm/src/main/java/org/apache/arrow/algorithm/deduplicate/VectorDeduplicator.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.algorithm.deduplicate;
+
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.util.DataSizeRoundingUtil;
+import org.apache.arrow.util.Preconditions;
+import org.apache.arrow.vector.BitVectorHelper;
+import org.apache.arrow.vector.IntVector;
+import org.apache.arrow.vector.ValueVector;
+
+import io.netty.buffer.ArrowBuf;
+
+/**
+ * Remove adjacent equal elements from a vector.
+ * If the vector is sorted, it removes all duplicated values in the vector.
+ * @param <V> vector type.
+ */
+public class VectorDeduplicator<V extends ValueVector> implements AutoCloseable {
+
+  /**
+   * Bit set for distinct values.
+   * If the value at some index is not equal to the previous value,
+   * its bit is set to 1, otherwise its bit is set to 0.
+   */
```

Comment:
this makes a speed-memory tradeoff, it seems like for all the operations defined on the method they would be much more efficient using a representation that stores the actual indices of the run starts.

Another possible option is to use a RoaringBitmap.

Reply:
This is a general problem. In the design, we have considered 3 different representations:

1. start indices represented as a bit set.
2. start indices represented as an byte vector.
3. an integer vector storing the start indices.

Generally speaking, 1 is cpu intensive, while 2 and 3 are memory intensive. 

We finally selected option 1, because in the extreme case (where each element is distinct from its previous one), the memory requirement of 1 is modest (in practice, the input vector can be large), while the start indices for 3 can be larger than the original vector (e.g. the original vector is a TinyIntVector).

Would you please give some suggestions?

Reply:
If the use-case is to apply this universally and memory is a concern then as you note this approach is reasonable.  As noted above, a [Roaring Bitmap](https://arxiv.org/pdf/1603.06549.pdf) implementation would potentially give you even a small set size and also potentially make computation faster, but this is more complicated and can potentially be done as a follow-up.

I'm still not clear on the exact use-case here, but if this were for RLE per-se, you could simply abort the computation if the first N elements were all set to one.  It really depends on the use-case, I figured it was worth a discussion in either case.

Reply:
Thanks for the recommended paper. It seems like something useful and interesting. Let's do it in a follow-up.

I agree with you that the choice depends on the specific use case. For our use case (big data shuffling and external sort), memory is the main concern. So we chose this design.

However, it may not work well for cases where cpu resources is the main concern. 

So what do you suggest for now? Do you think we should provide different implementations?

Reply:
I'll take another look at the implementation, but keeping as is should be OK for now.

On a general note it seems like you are adding a lot of algorithms that have potential design trade-offs.  If you know the algorithms you are going to add, I think it would be helpful to write-up a short document describing the algorithms you intend to add and share it on the mailing list, so feedback can be provided ahead of time (it doesn't need to be long, but should probably cover, at a high level any on-heap components, and processing vs memory trade-off for future algorithms).

Reply:
Thanks for the good suggestion. I will start more discussions in the mainling list.

Reply:
> Thanks for the good suggestion. I will start more discussions in the mainling list.

Thank you. Probably just one thread with a list of planned algorithms is enough (a google doc for commenting might be helpful).